src/reader/zarr_read.rs

```diff
@@ -184,30 +184,67 @@ impl ZarrProjection {
         }
     }
 
-    pub(crate) fn update(&mut self, other_proj: ZarrProjection) {
-        if other_proj.projection_type == ProjectionType::Null {
-            return;
-        }
-
-        if self.projection_type == ProjectionType::Null {
-            self.projection_type = other_proj.projection_type;
-            self.col_names = other_proj.col_names;
-            return;
-        }
-
-        let col_names = self.col_names.as_mut().unwrap();
-        if other_proj.projection_type == self.projection_type {
-            let mut s: HashSet<String> = HashSet::from_iter(col_names.clone());
-            let other_cols = other_proj.col_names.unwrap();
-            s.extend::<HashSet<String>>(HashSet::from_iter(other_cols));
-            self.col_names = Some(s.into_iter().collect_vec());
-        } else {
-            for col in other_proj.col_names.as_ref().unwrap() {
-                if let Some(index) = col_names.iter().position(|value| value == col) {
-                    col_names.remove(index);
+    pub(crate) fn update(&mut self, other_proj: ZarrProjection) -> ZarrResult<()> {
+        match (&self.projection_type, &other_proj.projection_type) {
+            (_, ProjectionType::Null) => (),
+            (ProjectionType::Null, _) => {
+                self.projection_type = other_proj.projection_type;
+                self.col_names = other_proj.col_names;
+                self.col_indices = other_proj.col_indices;
+            }
+            (ProjectionType::SelectByIndex, ProjectionType::SelectByIndex) => {
+                let mut indices = self
+                    .col_indices
+                    .take()
+                    .expect("ZarrProjection missing indices");
```

Comment:
```suggestion
                    .ok_or("ZarrProjection missing indices")?;
```

Could probably be done other places as well.

Reply:
changed it everywhere in the function.